source/mesh_deformation/fastscape.cc

```diff
@@ -0,0 +1,1632 @@
+                                                                    velocity_table);
+
+      VectorFunctionFromScalarFunctionObject<dim> vector_function_object(
+        [&](const Point<dim> &p) -> double
+      {
+        return velocities->value(p);
+      },
+      dim-1,
+      dim);
+
+      VectorTools::interpolate_boundary_values (mesh_deformation_dof_handler,
+                                                *boundary_ids.begin(),
+                                                vector_function_object,
+                                                mesh_velocity_constraints);
+    }
+
+
+    template <int dim>
+    std::vector<std::vector<double>>
+    FastScape<dim>::get_aspect_values() const
+    {
+
+      const types::boundary_id relevant_boundary = this->get_geometry_model().translate_symbolic_boundary_name_to_id ("top");
+      std::vector<std::vector<double>> temporary_variables(dim+2, std::vector<double>());
+
+      // Get a quadrature rule that exists only on the corners, and increase the refinement if specified.
+      const QIterated<dim-1> face_corners (QTrapez<1>(),
+                                           pow(2,additional_refinement+surface_refinement_difference));
+
+      FEFaceValues<dim> fe_face_values (this->get_mapping(),
+                                        this->get_fe(),
+                                        face_corners,
+                                        update_values |
+                                        update_quadrature_points);
+
+      typename DoFHandler<dim>::active_cell_iterator
+      cell = this->get_dof_handler().begin_active(),
+      endc = this->get_dof_handler().end();
+
+      for (; cell != endc; ++cell)
+        if (cell->is_locally_owned() && cell->at_boundary())
+          for (unsigned int face_no = 0; face_no < GeometryInfo<dim>::faces_per_cell; ++face_no)
+            if (cell->face(face_no)->at_boundary())
+              {
+                if ( cell->face(face_no)->boundary_id() != relevant_boundary)
+                  continue;
+
+                std::vector<Tensor<1,dim>> vel(face_corners.size());
+                fe_face_values.reinit(cell, face_no);
+                fe_face_values[this->introspection().extractors.velocities].get_function_values(this->get_solution(), vel);
+
+                for (unsigned int corner = 0; corner < face_corners.size(); ++corner)
+                  {
+                    const Point<dim> vertex = fe_face_values.quadrature_point(corner);
+
+                    // Find what x point we're at. Add 1 or 2 depending on if ghost nodes are used.
+                    const double indx = 1+use_ghost_nodes+(vertex(0) - grid_extent[0].first)/dx;
+
+                    // If our x or y index isn't close to a whole number, then it's likely an artifact
+                    // from using an over-resolved quadrature rule, in that case ignore it.
+                    if (abs(indx - round(indx)) >= precision)
+                      continue;
```

Comment:
Can you elaborate a bit on this? Are you saying the accuracy of the quadrature point position would be too low? That seems unlikely. Or are you saying someone could choose a different refinement level for the quadrature rule than the Fastscape mesh? This would look like an error to me that should not happen, so we should assert it does not happen (not ignore the false points).

Also if we ignore these points, wouldnt we have gaps in the data arrays we are filling here?

Reply:
This was done for two reasons:

1. Having multiple resolutions at the surface. Right now, a single quadrature rule exists for all the cells that makes sure to make enough interpolation points that the lowest resolution expected at the surface has enough points to fill the fastscape mesh. However, because this same rule is used on higher resolution cells, there will be extra points that do not correspond to the fastscape mesh, and this should skip over them. Is there perhaps a more straight forward way to do this?

3. The precision variable is more for if fastscape is used with a normal projection free surface, in which case the surface points likely won't match up exactly with the fixed fastscape mesh, and the quickest solution was to find a point that was near enough. Though maybe it would be better to not use the normal projection with fastscape and avoid this altogether.